### ISSUE TITLE
Certificate person grouping does not take trailing whitespace into account

### ISSUE THREAD

**Issue by ThiefMaster** [corona-warn-app/cwa-documentation#770]
## Avoid duplicates
* [x] Bug is not mentioned in the [FAQ](https://www.coronawarn.app/en/faq/)
* [x] Bug affects both Android **and** iOS, for specific issues / questions that apply only to one operating system, please raise them in the respective repositories:
    * [Android repository](https://github.com/corona-warn-app/cwa-app-android)
    * [iOS repository](https://github.com/corona-warn-app/cwa-app-ios)
* [x] Bug is not already reported in another issue

I do not know if this is a problem on iOS as well but I assume it's the case, so I'm opening the issue here.

## Technical details

- Device name: Pixel 5
- iOS/Android version: Android 12
- App version: 2.14.1

## Describe the bug

A certificate with a trailing whitespace in the last name is not correctly grouped with another one for the same person that does not have the whitespace.

## Steps to reproduce the issue

Scan two certs for the same person where one of them has a trailing whitespace in the name.

## Expected behaviour

Both certificates grouped together as they are for the same person

## Possible Fix

https://github.com/corona-warn-app/cwa-app-android/blob/main/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/common/certificate/CertificatePersonIdentifier.kt#L21-L22

Instead of just trimming the name, strip surrounding `<`  as well (since that's how spaces are written in the normalized name)

## Additional context

I got my two initial shots in Germany and a booster in France. Apparently the French system to generate QR codes does not strip surrounding whitespace and the person creating it accidentally entered a whitespace at the end of my last name so I ended up with this data in my 3/3 certificate (retrieved using the viewer at https://github.pathcheck.org/debug.html):

```json
"nam": {
  "fn": "MOEP ",
  "gn": "XXX YYY ADRIAN ZZZ",
  "fnt": "MOEP<",
  "gnt": "XXX<YYY<ADRIAN<ZZZ"
},
```

I replaced most of my actual names, but it still shows the problem clearly.

---
Internal Tracking ID: [EXPOSUREAPP-10868](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-10868)

**Comment by ThiefMaster** [comment]
The problem is that in the normalized name spaces are converted to `<` already during QR code generation. So if the non-normalized name has a trailing space, it becomes a trailing `<` in the actual data.

**Comment by Ein-Tim** [comment]
@ThiefMaster 

Understood, thank you for reporting & have a nice evening!

**Comment by mlenkeit** [comment]
@ThiefMaster thanks for bringing this up, valid bug report 👍

**Comment by larswmh** [comment]
Hi @ThiefMaster, thanks for your report. We have created an internal ticket and will raise this topic internally.
Internal Tracking ID: [EXPOSUREAPP-10868](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-10868)

---
Corona-Warn-App Open Source Team

**Comment by dsarkar** [comment]
FYI PR https://github.com/corona-warn-app/cwa-app-ios/pull/3979

**Comment by Ein-Tim** [comment]
Can you recheck this with 2.16 @ThiefMaster, please?

**Comment by ThiefMaster** [comment]
`Current Version 2.15.1` on Google Play... but I'll test once 2.16 is available!

**Comment by Ein-Tim** [comment]
@ThiefMaster Okay, under Android, 2.16.2 is in staged rollout at the moment. (-:

**Comment by ThiefMaster** [comment]
Still the same problem, even after fully deleting the latest certificate (also from recycle bin) and scanning it again. Without having seen the code in question, my guess would be that only literal spaces are stripped from the normalized name fields but NOT surrounding `<` characters

**Comment by vaubaehn** [comment]
@mlenkeit @thomasaugsten
Could you please cross-check whether this fix will break correct grouping of certificates with upcoming CCL when certificates are used with other applications?
It may be a minor problem for CWA, but a bigger problem for CovPassCheck or the DCC ticketing validation service.
Do I remember right that a similar trailing was also introduced to Android (ealier)?

/cc @oliver-steinbrecher

**Comment by MikeMcC399** [comment]
@ThiefMaster 
There are additional improvements in improving the tolerance for certificate matching included in 2.19 which is currently rolling out over the next 24 hours. See https://www.coronawarn.app/en/blog/2022-03-16-cwa-2-19/.

**Comment by MikeMcC399** [comment]
For CWA Android 2.19:
•	Fault tolerant name handling for certificates (EXPOSUREAPP-11617) https://github.com/corona-warn-app/cwa-app-android/pull/4858, https://github.com/corona-warn-app/cwa-app-android/pull/4872

The comments in [CertificatePersonIdentifier.kt](https://github.com/corona-warn-app/cwa-app-android/blob/release/2.19.x/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/common/certificate/CertificatePersonIdentifier.kt) give more detailed information:

```
    /**
     Method shall decide whether the DGCs belong to the same holder.
     Two DCCs shall be considered as belonging to the same holder, if:

     - the sanitized `dob` attributes are the same strings, and
     - one of:
     - the intersection/overlap of the name components of sanitized `a.nam.fnt` and `b.nam.fnt` has at least one
     element, and the intersection/overlap of the name components of sanitized `a.nam.gnt` and `b.nam.gnt` has at least
     one element or both are empty sets (`gnt` is an optional field)
     - the intersection/overlap of the name components of sanitized `a.nam.fnt` and `b.nam.gnt` has at least one
     element, and the intersection/overlap of the name components of sanitized `a.nam.gnt` and `b.nam.fnt` has at least
     one element
     */
```

and

```
/*
Function sanitize names according to the following requirements:

- dots `.` and dashes `-` shall be replaced by `<`
- the string shall be converted to upper-case
- German umlauts `Ä/ä`, `Ö/ö`, `Ü/ü` shall be replaced by `AE`, `OE`, `UE`
- German `ß` shall be replaced by `SS`
- the string shall be trimmed for leading and training whitespace
- the string shall be trimmed for leading and trailing `<`
- any whitespace in the string shall be replaced by `<`
- any occurrence of more than one `<` shall be replaced by a single `<`

Name fields shall be split into components as follows:

- the string shall be split by `<`
- components with the value `DR` shall be filtered out
 */
```

**Comment by MikeMcC399** [comment]
@ThiefMaster 

Will you check out if the new CWA Android 2.19.x solves the issue with trailing whitespace? In the meantime 2.19.1 has been rolled out following on from 2.19.0.

**Comment by MikeMcC399** [comment]
@dsarkar 
Please correct the label from [Fix 2.16](https://github.com/corona-warn-app/cwa-documentation/labels/Fix%202.16) to [Fix 2.19](https://github.com/corona-warn-app/cwa-documentation/labels/Fix%202.19).

**Comment by svengabr** [comment]
I have just checked the linked Jira issue.

Jira Ticket is flagged as:
Resolution: `Fixed`
Status: `Completed`
Fix Version/s: `Reelase 2.16`
